plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcCastPushdownTest.java

```diff
@@ -71,24 +71,55 @@ public void testJoinPushdownWithCast()
     @Test
     public void testCastFails()
     {
-        for (CastTestCase testCase : failCast()) {
-            assertThatThrownBy(() -> getQueryRunner().execute("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
-                    .hasMessageMatching("(.*)Cannot cast (.*) to (.*)");
+        Session withoutPushdown = Session.builder(getSession())
+                .setSystemProperty("allow_pushdown_into_connectors", "false")
+                .build();
+
+        for (FailCastTestCase testCase : failCast()) {
+            if (testCase.pushdownErrorMessage().isPresent()) {
+                assertThat(query("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
+                        .failure()
+                        .hasMessageMatching(testCase.pushdownErrorMessage().get());
+                assertThat(query(withoutPushdown, "SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
+                        .failure()
+                        .hasMessageMatching(testCase.errorMessage());
+            }
+            else {
+                assertThat(query("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
+                        .failure()
+                        .hasMessageMatching(testCase.errorMessage());
+            }
         }
     }
 
     public record CastTestCase(String sourceColumn, String castType, Optional<String> targetColumn)
     {
-        public CastTestCase(String sourceColumn, String castType)
+        public CastTestCase
         {
-            this(sourceColumn, castType, Optional.empty());
+            requireNonNull(sourceColumn, "sourceColumn is null");
+            requireNonNull(castType, "castType is null");
+            requireNonNull(targetColumn, "targetColumn is null");
         }
+    }
 
-        public CastTestCase
+    public record FailCastTestCase(String sourceColumn, String castType, String errorMessage, Optional<String> pushdownErrorMessage)
```

Comment:
Is there an example where we pass a valid pushdownError message ?

Reply:
For cases like:

```
assertThat(query(sessionWithoutPushdown, "SELECT id, CAST(c_decimal_25_2 AS TINYINT) from %s".formatted(table.getName())))
        .failure()
        .hasMessageContaining("Cannot cast '-129.49' to TINYINT");
assertThat(query("SELECT id, CAST(c_decimal_25_2 AS TINYINT) from %s WHERE id = 1".formatted(table.getName())))
        .failure()
        .hasMessageContaining("Bad value for type byte : -129");
```

can be converted to

```
new FailCastTestCase("c_decimal_25_2", "TINYINT", "Cannot cast '-129.49' to TINYINT", Optional.of("Bad value for type byte : -129"))
```

Reply:
In case of redshift we will have this case.
https://github.com/trinodb/trino/pull/22951/files#diff-2f548c43d7eac10982d24f0ceca0ab7799cbadefe6df9d85af168cb606c7753aR322-R327

Reply:
Let me remove this from this PR, as this is not needed with the current code base. I will add this constructor for the redshift case.

Reply:
FYI @SemionPar

Reply:
I think we can bring it closer to that PR - Here we don't use them.